core/src/engine/coro/pool.hpp

```diff
@@ -111,39 +126,74 @@ Pool<Task>::~Pool() = default;
 
 template <typename Task>
 typename Pool<Task>::CoroutinePtr Pool<Task>::GetCoroutine() {
-  struct CoroutineMover {
-    std::optional<Coroutine>& result;
-
-    CoroutineMover& operator=(Coroutine&& coro) {
-      result.emplace(std::move(coro));
-      return *this;
-    }
-  };
-
   std::optional<Coroutine> coroutine;
-  CoroutineMover mover{coroutine};
 
   // First try to dequeue from 'working set': if we can get a coroutine
   // from there we are happy, because we saved on minor-page-faulting (thus
   // increasing resident memory usage) a not-yet-de-virtualized coroutine stack.
-  if (used_coroutines_.try_dequeue(
-          GetUsedPoolToken<moodycamel::ConsumerToken>(), mover) ||
-      initial_coroutines_.try_dequeue(mover)) {
-    --idle_coroutines_num_;
+  if (local_coro_buffer<Task>.size() > 0) {
+    coroutine = std::move(local_coro_buffer<Task>.back());
+    local_coro_buffer<Task>.pop_back();
   } else {
-    coroutine.emplace(CreateCoroutine());
+    const std::size_t deque_num =
+        std::min(idle_coroutines_num_.load(std::memory_order_acquire),
+                 kLocalCoroutineMoveSize);
+    if (deque_num > 0) {
+      local_coro_buffer<Task>.resize(deque_num);
+      const std::size_t dequed_num = used_coroutines_.try_dequeue_bulk(
+          GetUsedPoolToken<moodycamel::ConsumerToken>(),
+          local_coro_buffer<Task>.data(), deque_num);
+      local_coro_buffer<Task>.resize(dequed_num);
+
+      if (dequed_num > 0) {
+        coroutine = std::move(local_coro_buffer<Task>.back());
+        local_coro_buffer<Task>.pop_back();
+        idle_coroutines_num_.fetch_sub(dequed_num, std::memory_order_release);
```

Comment:
Can we just replace `idle_coroutines_num_` with `used_coroutines_.size_approx()`? Maybe it requires accounting for `initial_coroutines_` as well?

Reply:
In this case, you need to check the size of Initial_coroutines_ . Doesn't look like it will be faster. Provided that size_approx goes through the queues of all producers to calculate the size.